### ISSUE TITLE
Question: PromiseKit 4.x and promise reuse (and checking if existing promise is pending)

### ISSUE THREAD

**Issue by ryanschneider** [mxcl/PromiseKit#555]
* PromiseKit version: 4.0.1
* Xcode version: 8.1 beta

I've got a question regarding PromiseKit 4.0.   In previous versions I used a pattern where I would piggy back off an existing Promise if there was one available and pending, like so:

```swift
func doSomethingExpensive() -> Promise<Void> {
    if let existingPromise = self._expensivePromise where existingPromise.pending == true {
        return existingPromise
    }
    // otherwise, create and return the promise
    self. _expensivePromise = firstly {
        //...
    }
    .then {
        //...
    }
    .always {
        self._uploadPromise = nil
    }

    return self. _expensivePromise
```

But in PromiseKit 4.0, `.state` isn't public.  I'm guessing that means you don't want outside code inspecting the pending state of a Promise?  Was that always the intent, and now it's just more explicit?  Or am I possibly missing an new API (I saw `.tap()` but don't quite grok it)?

More generally, is there a preferred pattern for reusing "in-flight" promises?  

I know technically I don't need the `where existingPromise.pending == true` clause, since I set the promise to `nil` upon completion in the `always` clause, but relying fully on that feels a little too implicit and hard to follow.

FWIW, in googling around, I realized I asked a very similar question when I first started learning PromiseKit earlier this year:

http://stackoverflow.com/questions/36607559/whats-the-best-swift-pattern-for-returning-a-pending-promisekit-promise-if-one

Ideally, I'd still be interested in what an ideal `makeOrReusePromise()` type function would look like, in your opinion.

**Comment by mxcl** [comment]
`pending` is now `isPending` per Swift 3 API guidelines.

**Comment by ryanschneider** [comment]
Ah, I totally missed the `Promise+Properties.swift` file when looking through the code. :)

**Comment by mxcl** [comment]
The pattern you're using is what I'd do. I'd probably make it `lazy` and not optional though.
